tests/backup/backup_upgrade_test.go

```diff
@@ -204,3 +205,366 @@ var _ = Describe("{StorkUpgradeWithBackup}", func() {
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			preRuleNames = make(map[string]string)
+			preRuleUids = make(map[string]string)
+			log.InfoD("Creating pre backup rules for applications %v", Inst().AppList)
+			for _, appName := range Inst().AppList {
+				log.Infof("Creating pre backup rule for application [%s]", appName)
+				_, preRuleName, err := Inst().Backup.CreateRuleForBackup(appName, orgID, "pre")
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of pre-backup rule for application [%s]", appName))
+				preRuleUid := ""
+				if preRuleName != "" {
+					preRuleUid, err = Inst().Backup.GetRuleUid(orgID, ctx, preRuleName)
+					log.FailOnError(err, "Fetching pre backup rule [%s] uid", preRuleName)
+					log.Infof("Pre backup rule [%s] uid: [%s]", preRuleName, preRuleUid)
+				}
+				for i := 0; i < len(appNamespaces[appName]); i++ {
+					preRuleNames[appName] = preRuleName
+					preRuleUids[appName] = preRuleUid
+				}
+			}
+			postRuleNames = make(map[string]string)
+			postRuleUids = make(map[string]string)
+			log.InfoD("Creating post-backup rules for applications %v", Inst().AppList)
+			for _, appName := range Inst().AppList {
+				log.Infof("Creating post backup rule for application [%s]", appName)
+				_, postRuleName, err := Inst().Backup.CreateRuleForBackup(appName, orgID, "post")
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of post-backup rule for application [%s]", appName))
+				postRuleUid := ""
+				if postRuleName != "" {
+					postRuleUid, err = Inst().Backup.GetRuleUid(orgID, ctx, postRuleName)
+					log.FailOnError(err, "Fetching post backup rule [%s] uid", postRuleName)
+					log.Infof("Post backup rule [%s] uid: [%s]", postRuleName, postRuleUid)
+				}
+				for i := 0; i < len(appNamespaces[appName]); i++ {
+					postRuleNames[appName] = postRuleName
+					postRuleUids[appName] = postRuleUid
+				}
+			}
+		})
+		Step("Create backups with and without pre- and post-backup rules", func() {
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			createBackupWithRulesTask := func(appName string) {
+				namespace := appNamespaces[appName][0]
+				backupName := fmt.Sprintf("%s-%s-%v-with-rules", BackupNamePrefix, namespace, time.Now().Unix())
+				labelSelectors := make(map[string]string, 0)
+				log.InfoD("Creating a backup of namespace [%s] with pre- and post-backup rules", namespace)
+				err := CreateBackup(backupName, SourceClusterName, backupLocationName, backupLocationUid, []string{namespace},
+					labelSelectors, orgID, srcClusterUid, preRuleNames[appName], preRuleUids[appName], postRuleNames[appName], postRuleUids[appName], ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of backup [%s]", backupName))
+				backupWithRuleNames = SafeAppend(&mutex, backupWithRuleNames, backupName).([]string)
+			}
```

Comment:
use separate mutex for `backupWithRuleNames`, as lock should "belong" to the object that it is protecting.